service/src/java/org/apache/hive/service/server/HiveServer2.java

```diff
@@ -1201,63 +1204,64 @@ private static void startHiveServer2() throws Throwable {
     }
   }
 
-  private void maybeStartCompactorThreads(HiveConf hiveConf) throws Exception {
+  public Map<String, Integer> maybeStartCompactorThreads(HiveConf hiveConf) {
+    Map<String, Integer> startedWorkers = new HashMap<>();
     if (MetastoreConf.getVar(hiveConf, MetastoreConf.ConfVars.HIVE_METASTORE_RUNWORKER_IN).equals("hs2")) {
-      int numWorkers = MetastoreConf.getIntVar(hiveConf, MetastoreConf.ConfVars.COMPACTOR_WORKER_THREADS);
-      List<Map.Entry<String, String>> entries = hiveConf.getMatchingEntries(Constants.COMPACTION_POOLS_PATTERN);
+      Ref<Integer> numWorkers = new Ref<>(MetastoreConf.getIntVar(hiveConf, MetastoreConf.ConfVars.COMPACTOR_WORKER_THREADS));
+      Map<String, Integer> customPools = CompactorUtil.getPoolConf(hiveConf);
 
       StringBuilder sb = new StringBuilder(2048);
       sb.append("This HS2 instance will act as Compactor with the following worker pool configuration:\n");
-      sb.append("Global pool size: ").append(numWorkers).append("\n");
-
-      LOG.info("Initializing the compaction pools with using the global worker limit: {} ", numWorkers);
-      while (numWorkers > 0 && entries.size() > 0) {
-        Map.Entry<String, String> entry = entries.remove(0);
-        String poolName = entry.getValue();
-        int poolWorkers = hiveConf.getInt(entry.getKey(), 0);
+      sb.append("Global pool size: ").append(numWorkers.value).append("\n");
 
+      LOG.info("Initializing the compaction pools with using the global worker limit: {} ", numWorkers.value);
+      customPools.forEach((poolName, poolWorkers) -> {
         if (poolWorkers == 0) {
           LOG.warn("Compaction pool ({}) configured with zero workers. Skipping pool initialization", poolName);
           sb.append("Pool not initialized, 0 size: ").append(poolName).append("\n");
-          continue;
         }
-        if (poolWorkers > numWorkers) {
-          LOG.warn("Global worker pool exhausted, compaction pool ({}) will be configured with less workers than the " +
-              "required number. ({} -> {})", poolName, poolWorkers, numWorkers);
-          poolWorkers = numWorkers;
+        else if (numWorkers.value == 0) {
+          LOG.warn("There are no available workers for the following compaction pool: {} ", poolName);
+          sb.append("Pool not initialized, no remaining free workers: ").append(poolName).append("\n");
```

Comment:
i would just log this line instead of the prev one. Als,o why do we need to add this into `sb`?

Reply:
This was added into sb in the existing code.
I removed adding to sb and logged warning instead.

Reply:
Yes, I just pushed new commit